llap-tez/src/java/org/apache/hadoop/hive/llap/tezplugins/LlapTaskSchedulerService.java

```diff
@@ -3154,4 +3165,28 @@ public void taskInfoUpdated(TezTaskAttemptID attemptId, boolean isGuaranteed) {
         + attemptId + ", " + newState);
     sendUpdateMessageAsync(ti, newState);
   }
+
+  private void updateMetrics(TaskAttemptImpl taskAttempt) {
+    // Only do it for map tasks
+    if (!isMapTask(taskAttempt)) {
+      return;
+    }
+    // Check if this task was already assigned to a node
+    NodeInfo nodeInfo = knownTasks.get(taskAttempt).assignedNode;
+    if (nodeInfo == null) {
+      return;
+    }
+
+    metrics.addTaskLatency(nodeInfo.shortStringBase, taskAttempt.getFinishTime() - taskAttempt.getLaunchTime());
+  }
+
+  private boolean isMapTask(TaskAttemptImpl taskAttempt) {
+    boolean isMapTask = false;
+    for(TezCounter counter : taskAttempt.getCounters().getGroup("HIVE")) {
```

Comment:
I am sure, there are more efficient ways to figure out, if this is a Map task. The CounterGroupBase for example has a findCounter, therefore iterating all counters of a group to find a specific one, is not necessary. Maybe, instead of looking at the counters, there are other indicators (i.e. at the attached Vertex), giving away if this a Map task.

Reply:
Accessing vertex data seems even more complicated.
Optimized the counter based solution as suggested.